### PR TITLE
Raw sql

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes

Includes commit from #5122

### CODE REVIEW

**Comment by trasher** [comment]
Maybe the use of QueryExpression in ORDER clause may target 9.4/bf

File: inc/migration.class.php [hunk 1]
```diff
@@ -393,7 +393,7 @@ function addField($table, $field, $type, $options = []) {
                $query = "UPDATE `$table`
                         SET `$field` = ".$params['update']." ".
                         $params['condition']."";
-               $DB->queryOrDie($query, $this->version." set $field in $table");
+               $DB->rawQueryOrDie($query, $this->version." set $field in $table");
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/migration.class.php]: Indeed, I think I can change :)

File: inc/dbmysqliterator.class.php [hunk 2]
```diff
@@ -379,22 +377,30 @@ function buildQuery ($table, $crit = "", $log = false) {
...
-      foreach ($clause as $o) {
-         $fields = explode(',', $o);
-         foreach ($fields as $field) {
-            $new = '';
-            $tmp = explode(' ', trim($field));
-            $new .= DBmysql::quoteName($tmp[0]);
-            // ASC OR DESC added
-            if (isset($tmp[1]) && in_array($tmp[1], ['ASC', 'DESC'])) {
-               $new .= ' '.$tmp[1];
+      if ($clause instanceof \QueryExpression) {
```
Comment (cedric-anne) [inline inc/dbmysqliterator.class.php]: You should add `QueryExpression` as possible type for `$clause` param in doc block.

File: inc/dbmysql.class.php [hunk 3]
```diff
@@ -196,14 +217,31 @@ function query($query) {
...
-      $res = $this->query($query);
+      Toolbox::deprecated();
+      return $this->rawQueryOrDie($query, $message);
+   }
+
+   /**
+    * Execute a MySQL query and die
+    * (optionnaly with a message) if it fails
+    *
+    * @since 0.84
```
Comment (cedric-anne) [inline inc/dbmysql.class.php]:
```suggestion
    * @since 10.0.0
```

File: inc/ipnetwork.class.php [hunk 4]
```diff
@@ -543,9 +541,10 @@ static function searchNetworks($relation, $condition, $entityID = -1, $recursive
 
          if ($relation == "equals") {
             for ($i = $startIndex; $i < 4; ++$i) {
-               $WHERE .= " AND (`".$addressDB[$i]."` & '".$netmaskPa[$i]."')=
-                               ('".$addressPa[$i]."' & '".$netmaskPa[$i]."')
-                           AND ('".$netmaskPa[$i]."' = `".$netmaskDB[$i]."`)";
+               $WHERE = [
+                  new \QueryExpression("(`".$addressDB[$i] . "` & '" . $netmaskPa[$i] . "') = ('".$addressPa[$i]."' & '".$netmaskPa[$i]."')"),
+                  $netmaskPa[$i] => $netmaskDB[$i]
```
Comment (cedric-anne) [inline inc/ipnetwork.class.php]:
```suggestion
                  new \QueryExpression("'$netmaskPa[$i]' = '$netmaskDB[$i]'"),
```

**Comment by cedric-anne** [comment]
You hsould add changelog entries too.

File: inc/migration.class.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (trasher) [reply inc/migration.class.php]: It's a quite big change, will do in another PR

**Pull request merged**